src/core/server/saved_objects/migrationsv2/integration_tests/multiple_kibana_nodes.test.ts

```diff
@@ -0,0 +1,219 @@
+}
+
+function createRoot({ logFileName }: CreateRootConfig) {
+  return kbnTestServer.createRoot(
+    {
+      elasticsearch: {
+        hosts: [esTestConfig.getUrl()],
+        username: kibanaServerTestUser.username,
+        password: kibanaServerTestUser.password,
+      },
+      migrations: {
+        skip: false,
+        enableV2: true,
+        batchSize: 100, // fixture contains 5000 docs
+      },
+      logging: {
+        appenders: {
+          file: {
+            type: 'file',
+            fileName: logFileName,
+            layout: {
+              type: 'pattern',
+            },
+          },
+        },
+        loggers: [
+          {
+            name: 'root',
+            appenders: ['file'],
+          },
+          {
+            name: 'savedobjects-service',
+            appenders: ['file'],
+            level: 'debug',
+          },
+        ],
+      },
+    },
+    {
+      oss: true,
+    }
+  );
+}
+
+describe('migration v2', () => {
+  let esServer: kbnTestServer.TestElasticsearchUtils;
+  let rootA: Root;
+  let rootB: Root;
+  let rootC: Root;
+
+  const migratedIndex = `.kibana_${pkg.version}_001`;
+  const fooType: SavedObjectsType = {
+    name: 'foo',
+    hidden: false,
+    mappings: { properties: { status: { type: 'text' } } },
+    namespaceType: 'agnostic',
+    migrations: {
+      '7.14.0': (doc) => {
+        if (doc.attributes?.status) {
+          doc.attributes.status = doc.attributes.status.replace('not_migrated', 'migrated');
+        }
+        return doc;
+      },
+    },
+  };
+
+  beforeAll(async () => {
+    await removeLogFiles();
+  });
+
+  afterAll(async () => {
+    await new Promise((resolve) => setTimeout(resolve, 10000));
+  });
+
+  beforeEach(async () => {
+    rootA = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_A.log`),
+    });
+    rootB = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_B.log`),
+    });
+    rootC = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_C.log`),
+    });
+
+    const { startES } = kbnTestServer.createTestServers({
+      adjustTimeout: (t: number) => jest.setTimeout(t),
+      settings: {
+        es: {
+          license: 'trial',
+          // original SO:
+          // [
+          //   { id: 'foo:1', type: 'foo', foo: { status: 'not_migrated_1' } },
+          //   { id: 'foo:2', type: 'foo', foo: { status: 'not_migrated_2' } },
+          //   { id: 'foo:3', type: 'foo', foo: { status: 'not_migrated_3' } },
+          // ];
+          dataArchive: Path.join(__dirname, 'archives', '7.13.0_5k_so.zip'),
+        },
+      },
+    });
+    esServer = await startES();
+  });
+
+  afterEach(async () => {
+    await Promise.all([rootA.shutdown(), rootB.shutdown(), rootC.shutdown()]);
+
+    if (esServer) {
+      await esServer.stop();
+    }
+  });
+
+  it('migrates saved objects normally when multiple Kibana instances are running', async () => {
```

Comment:
Open to suggestions of anything more I could try in this test to simulate race conditions. Right now it's basically just testing that the algorithm is idempotent if there are multiple Kibanas doing migrations.

Reply:
I don't think we're really going to be able to force a race here, this test simply serves as a smoke test to ensure that if we do make an incorrect change to the algorithm, we hopefully catch the race at least sometimes.

Have you been able to verify manually that this test will catch non-idempotentcy if we introduce it intentionally? For example, if you comment out the `conflicts: 'proceed'` line here, will this test fail? https://github.com/elastic/kibana/blob/6d48a50c2a4ecff9188e4d2c9f7d426ba4398294/src/core/server/saved_objects/migrationsv2/actions/index.ts#L411-L413

Reply:
> Have you been able to verify manually that this test will catch non-idempotentcy if we introduce it intentionally? For example, if you comment out the conflicts: 'proceed' line here, will this test fail?

Yep 🙂  
I tested manually by removing `conflicts: 'proceed'` in both places it is used in the algo (when picking up updated mappings and also when reindexing into the target index). In both cases it will cause the test to fail.